text/0000-permissions.md

```diff
@@ -0,0 +1,186 @@
+}
+```
+Dang it! But with permissions, we can easily fix this. So let us change some things in our `impl` of `Person`:
+```rust
+impl Person {
+  pub fn aging(&'Age mut self) { // this function needs the permission `'Age`
+    self.age += 1;
+  }
+  
+  pub fn rename(&'Name mut self, new_name: String) -> String { // this function needs the permission `'Name`
+    std::mem::replace(&mut self.name, new_name)
+  }
+}
+```
+Now let us change the borrow for `Department` and `Calendar`:
+```rust
+pub struct Department<'p>(&'p 'Name mut Person); // the borrow has the permission `'Name`
+pub struct Calendar<'p>(&'p 'Age mut Person); // the borrow has the permission `'Age`
+```
+And now let's see our main code:
+```rust
+fn main() {
+  let mut person = Person {
+    name: "Mike",
+    age: 0,
+  };
+  
+  let department = Department(&mut person); // implicit permission
+  let calendar = Calendar(&'Age mut person); // explicit permission
+  
+  calendar.next_year();
+  calendar.next_year();
+  calendar.next_year();
+  department.rename("Jonathan");
+  calendar.next_year();
+}
+```
+And TADA: our code compiles and works 🥳.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The compiler should know what the permissions allow of data to use.
+For example, when two functions need the same data mutably and normally it wouldn't compile, they shouldn't just add two different permissions to evade it:
+```rust
+impl Hack {
+  pub fn 'HackA hack_a(&mut self) {
+    self.one_data./*...*/
+  }
+  
+  pub fn 'HackB hack_b(&mut self) {
+    self.one_data./*...*/
+  }
+}
+```
+Instead, the compiler should notify that the permissions intercept with each other and they should create a sub-permission for both:
+```rust
+impl Hack {
+  pub fn 'OneData + 'HackA hack_a(&mut self)
+  where
+    self.one_data: 'OneData // I'm unsure about this syntax
+  {
+    self.one_data./*...*/
+  }
+  
+  pub fn 'OneData + 'HackB hack_b(&mut self)
+  where
+    self.one_data: 'OneData
+  {
+    self.one_data./*...*/
+  }
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It may be confusing especially for newbies to differentiate between lifetimes and permissions. Maybe something else than `'`?
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+I think this is one of the cleanest options for these problems. Other devs may be changing their whole code structure for that or use `unsafe`, but I like the idea of permissions.
+It maybe "could" be done with macros, but then it would also generate unsafe code and wont be 100% safe.. but the simplest impl of permissions may be possible to implement via them.
+
+# Prior art
+[prior-art]: #prior-art
+
+I don't know any language who has done something like this before.
+The idea came to me by minecraft server development and there its used for giving players the permission to execute commands, so it may be similar in the smallest way possible 😅.
```

Comment:
```suggestion
The idea came to me by Minecraft server development, where it is used for giving players the permission to execute commands, so it may be similar in the smallest way possible 😅.
```